client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java

```diff
@@ -66,26 +75,29 @@ public PositionImpl(Map<SegmentWithRange, Long> segments) {
 
     @Override
     public Set<Segment> getOwnedSegments() {
+        applySegmentOffsetUpdatesIfNeeded();
         return Collections.unmodifiableSet(ownedSegments.keySet());
     }
 
     @Override
     public Map<Segment, Long> getOwnedSegmentsWithOffsets() {
+        applySegmentOffsetUpdatesIfNeeded();
         return Collections.unmodifiableMap(ownedSegments);
     }
     
     @Override
     Map<SegmentWithRange, Long> getOwnedSegmentRangesWithOffsets() {
+        applySegmentOffsetUpdatesIfNeeded();
         HashMap<SegmentWithRange, Long> result = new HashMap<>();
         for (Entry<Segment, Long> entry : ownedSegments.entrySet()) {
```

Comment:
ownedSegments can be modified by the different thread while iterating here.

Reply:
Correct, good catch. Now, when `applySegmentOffsetUpdatesIfNeeded()` is executed, the `PositionImpl` object copies the `ownedSegments` map and only then sets the right offsets representing the point at which the event was read.